### ISSUE TITLE
How to do Lora Training correctly?

### ISSUE THREAD

**Issue by petterthowsen** [oobabooga/text-generation-webui#2608]
**Description**

I tried training a lora for [Wizard-Vicuna-13B-16F](https://huggingface.co/TheBloke/Wizard-Vicuna-13B-Uncensored-HF)

I made a text file like with a bunch of examples, separated by \n\n\n

Each example is an example of running a command on behalf of the user.

I tried a bunch of times with varying settings. 10 epochs, 20 epochs. Lora rank 128 or 256 and the alpha double of that. I also tried a learning rate of 3e-3 instead of 3e-4.

Usually when trying the lora with an input like 'can you start spotify?' it would respond with something like 'sure, whats the command for that?'. Other times, it would go into some wierd nonesense response like 'yes yes yes {€{€$€$€$€€$€€$€€$€'.

Not sure what i'm doing wrong here. Is the dataset too small? Or am I making a technical mistake in formatting the data? Do I need each example to include my context prompt perhaps?

Thanks in advance.

**Additional Context**

The training file looks like this:
USER: Can you open Spotify for me?
ASSISTANT: Absolutely. [command=terminal "start spotify"]

USER: Could you please show me the content of the current directory?
ASSISTANT: Of course. [command=terminal "ls"]

USER: Could you search the web for the term 'machine learning'?
ASSISTANT: Sure, I can do that. [command=web_search "machine learning"]

USER: Please browse the OpenAI website.
ASSISTANT: Alright. [command=web_browse "https://www.openai.com"]

USER: Can you show me the current system status?
ASSISTANT: Sure. [command=terminal "top"]

USER: Could you open the calculator?
ASSISTANT: Absolutely. [command=terminal "open calculator"]

USER: Please search the web for 'the best pizza recipe'.
ASSISTANT: Alright. [command=web_search "the best pizza recipe"]

USER: Can you check my internet connection?
ASSISTANT: Sure. [command=terminal "ping google.com"]

USER: Please open the Paint program.
ASSISTANT: Sure. [command=terminal "start mspaint"]

USER: Can you show me the IP configuration?
ASSISTANT: Of course. [command=terminal "ipconfig"]

USER: Could you search the web for the nearest pizza place?
ASSISTANT: Sure, I can do that. [command=web_search "nearest pizza place"]

USER: Please browse the Wikipedia page for 'Artificial Intelligence'.
ASSISTANT: Absolutely. [command=web_browse "https://en.wikipedia.org/wiki/Artificial_intelligence"]

USER: Can you open the Terminal?
ASSISTANT: Sure. [command=terminal "open Terminal"]

USER: Could you show me the disk usage of my computer?
ASSISTANT: Of course. [command=terminal "df -h"]

USER: Please search the web for 'how to train a neural network'.
ASSISTANT: Sure. [command=web_search "how to train a neural network"]

USER: Can you browse the URL 'www.github.com'?
ASSISTANT: Absolutely. [command=web_browse "www.github.com"]

USER: Can you check the current date and time?
ASSISTANT: Of course. [command=terminal "date"]

USER: Please show me the running processes.
ASSISTANT: Absolutely. [command=terminal "ps aux"]

USER: Could you search the web for 'Python programming tutorials'?
ASSISTANT: Sure, I can do that. [command=web_search "Python programming tutorials"]

USER: Please browse the New York Times website.
ASSISTANT: Alright. [command=web_browse "https://www.nytimes.com"]

USER: Can you open Microsoft Word?
ASSISTANT: Sure. [command=terminal "start winword"]

USER: Could you show me the network statistics?
ASSISTANT: Of course. [command=terminal "netstat"]

USER: Please search the web for 'latest news on artificial intelligence'.
ASSISTANT: Sure. [command=web_search "latest news on artificial intelligence"]

USER: Can you browse the URL www.medium.com?
ASSISTANT: Absolutely. [command=web_browse "www.medium.com"]

USER: Can you open the command prompt?
ASSISTANT: Sure. [command=terminal "start cmd"]

**Comment by Ph0rk0z** [comment]
If this is your dataset, it's like an eye dropper going into an ocean. Did you try running more epochs? What was your final result numbers wise.

**Comment by petterthowsen** [comment]
@Ph0rk0z I didn't get any numbers or anything reported in the console. It just says training started and after a while it says training finished. I'm running it on runpod, so I attach the ssh session to the stdout of the oobabooga process using tail -f /proc/<pid>/fd/1. I tried 8, 12, 20 epochs.

How could I make this data into JSON format? It's a bit confusing :\

**Comment by petterthowsen** [comment]
Suppose I have 1000 examples in the dataset, as opposed to ~30. How many epochs should I try with? 20? 100? 1000?

**Comment by Ph0rk0z** [comment]
To make it into json copy alapaca and use that. The actual formatting part should be done with a script but we're a bit short on data processing ones.

Try 20 and see what your loss values are. 1.5-1.0 is good for style but in your case I think below 1 is going to be better.since you have so few examples. On a normal size data set this number is like 1-6.Usually people do like 2 or 3 and it learns.

**Comment by petterthowsen** [comment]
Where can I see the loss values though? I'm attaching to the process by using `tail -f /proc//fd/1` but it's only showing "Training started" then "Training finished".

**Comment by petterthowsen** [comment]
@tensiondriven I appreciate the feedback but I'm not sure if this is how I want to do it. Once I get the hang of how I can effectively train loras, I'll transition to using bigger models (30b first and then probably 65b when it comes out).

In my prototype using Open AI's GPT 3.5, I am doing something similar to what you're suggesting, but for a different purpose.

Anyway, I'd like to learn how to make loras anyway.
